wicketd/src/artifacts/update_plan.rs

```diff
@@ -0,0 +1,1016 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Constructor for the `UpdatePlan` wicketd uses to drive sled mupdates.
+//!
+//! This is a "plan" in name only: it is a strict list of which artifacts to
+//! apply to which components; the ordering and application of the plan lives
+//! elsewhere.
+
+use super::extracted_artifacts::ExtractedArtifacts;
+use super::extracted_artifacts::HashingNamedUtf8TempFile;
+use super::ArtifactIdData;
+use super::Board;
+use super::ExtractedArtifactDataHandle;
+use super::RepositoryError;
+use super::UpdatePlan;
+use anyhow::anyhow;
+use hubtools::RawHubrisArchive;
+use omicron_common::api::external::SemverVersion;
+use omicron_common::api::internal::nexus::KnownArtifactKind;
+use omicron_common::update::ArtifactHash;
+use omicron_common::update::ArtifactHashId;
+use omicron_common::update::ArtifactId;
+use omicron_common::update::ArtifactKind;
+use slog::info;
+use slog::Logger;
+use std::collections::btree_map;
+use std::collections::BTreeMap;
+use std::collections::HashMap;
+use std::io;
+use std::io::Read;
+use tufaceous_lib::HostPhaseImages;
+use tufaceous_lib::RotArchives;
+
+/// `UpdatePlanBuilder` mirrors all the fields of `UpdatePlan`, but they're all
+/// optional: it can be filled in as we read a TUF repository.
+/// [`UpdatePlanBuilder::build()`] will (fallibly) convert from the builder to
+/// the final plan.
+#[derive(Debug)]
+pub(super) struct UpdatePlanBuilder<'a> {
+    // fields that mirror `UpdatePlan`
+    system_version: SemverVersion,
+    gimlet_sp: BTreeMap<Board, ArtifactIdData>,
+    gimlet_rot_a: Option<ArtifactIdData>,
+    gimlet_rot_b: Option<ArtifactIdData>,
+    psc_sp: BTreeMap<Board, ArtifactIdData>,
+    psc_rot_a: Option<ArtifactIdData>,
+    psc_rot_b: Option<ArtifactIdData>,
+    sidecar_sp: BTreeMap<Board, ArtifactIdData>,
+    sidecar_rot_a: Option<ArtifactIdData>,
+    sidecar_rot_b: Option<ArtifactIdData>,
+    host_phase_1: Option<ArtifactIdData>,
+    trampoline_phase_1: Option<ArtifactIdData>,
+    trampoline_phase_2: Option<ArtifactIdData>,
+    host_phase_2_hash: Option<ArtifactHash>,
+    control_plane_hash: Option<ArtifactHash>,
```

Comment:
Why store `trampoline_phase_2` (i.e. not just a hash) and `host_phase_2_hash` here? Would be worth a comment.

Reply:
Added a comment in c27dc85d: we have to send the trampoline phase 2 to MGS, but installinator fetches the host phase 2 from us over the bootstrap network (we just have to send it the hash via MGS).